torch/_dynamo/variables/builder.py

```diff
@@ -2479,132 +2676,42 @@ def _automatic_dynamic(
-                            if dim is not None and stride[i] != dim:
-                                log.debug(
-                                    "automatic dynamic %s stride(%s) %s != %s",
-                                    name,
-                                    i,
-                                    stride[i],
-                                    dim,
-                                )
-                                get_chromium_event_logger().log_instant_event(
-                                    "automatic_dynamic",
-                                    time.time_ns(),
-                                    {
-                                        "name": name,
-                                        "dim_changed": i,
-                                        "reason": "stride change",
-                                        "cached": str(dim),
-                                        "new": str(stride[i]),
-                                    },
-                                )
-                                frame_state_entry.stride[i] = None
-        tx.output.frame_state[name] = frame_state_entry
 
-    if (st := tx.distributed_state) is None:
-        stride = e.stride() if not is_sparse_any(e) else ()
-        update_frame_state(e.size(), stride)
-        frame_state_entry = tx.output.frame_state[name]
-    elif st.all_states is None:
-        # Preflight, always pretend as if it's static
-        frame_state_entry = FrameStateSizeEntry(
-            size=e.size(), scalar=None, stride=e.stride()
-        )
-        st.local_state.input_sizes[name] = list(e.size())
-        st.local_state.input_strides[name] = list(e.stride())
-    else:
-        # Apply the updates
-        for sub_state in st.all_states:
-            # Not all inputs are necessarily present on all ranks
-            if name in sub_state.input_sizes and name in sub_state.input_strides:
-                update_frame_state(
-                    sub_state.input_sizes[name], sub_state.input_strides[name]
+    # This mimics stride inference algorithm in _create_symbolic_sizes_strides_storage_offset
+    ex_size = e.size()
+    if not is_sparse_any(e):
+        ex_stride = e.stride()
+        dim = e.dim()
+
+        stride = [None] * dim
+        while any(x is None for x in stride):
+            candidates = {
+                ex_size[i] * ex_stride[i]: InferStride(i)
+                for i in range(dim)
+                if stride[i] is not None and ex_stride[i] >= 0
+            }
+            val_list = sorted(
+                [(ex_stride[i], i) for i in range(dim) if stride[i] is None],
+                key=_nested_int_aware_sort,
+            )
+            for _, i in val_list:
+                if stride[i] is None and ex_stride[i] in candidates:
+                    stride[i] = candidates[ex_stride[i]]
+                    candidates[ex_stride[i] * ex_size[i]] = InferStride(i)
+
+            if any(x is None for x in stride):
```

Comment:
Maybe you only want to do this `if` if we didn't find a candidate above? (i.e. you either find a candidate OR you swap out the smallest unbound stride, not both in the same inner loop)?

I could be wrong but this whole algorithm feels like maybe it could be simpler using a single outer loop pulling from a minheap...

Reply:
So first, this is based off an existing an algo. But I do know why the algo is written this way, so let me explain.

The inference algorithm here is based off the idea that most of the time, dimensions are contiguous with each other. Once we find the dimension with the smallest stride, we can successively look for contiguous strides from candidates so we can build inferred stride values.

```
            val_list = sorted(
                [(ex_stride[i], i) for i in range(dim) if stride[i] is None],
                key=_nested_int_aware_sort,
            )
```

Ideally, at this point the loop terminates. However, it is possible to have some strides that are completely unrelated to all other inner dimensions we have already handled. So at this point, if we still have strides left, we get the smallest one and give it a fresh variable

```
            if any(x is None for x in stride):
                # bind the smallest unbound stride to a new variable
                val, i = min(
                    [(ex_stride[i], i) for i in range(dim) if stride[i] is None],
                    key=_nested_int_aware_sort,
                )
                stride[i] = val
```

Binding this fresh variable may now unlock *more* contiguous strides which we can infer in the normal way. So we want to go back to the beginning of the loop and do the process again.

As for min heap, the problem is that the "smallest unbound stride" may be smaller than some strides which you successfully processed from contiguity. For example, if I have tensor of size `[3, 3, 3, 3]`, my contiguous strides will be processed in order `[1, 3, 9, 27]`, so if I have `[1, 4, 3, 9]`, I will process 1, 3, 9, and only THEN will I allocate a new stride for 4.

Reply:
It's totally reasonable to leave this as-is because of wanting to match an existing algorithm. I just want to make sure we're talking about the same thing here. This is the minheap version I propose:
```
    ex_size = e.size()
    ex_stride = e.stride()
    dim = e.dim()

    stride = [None] * dim
    pending = [(_nested_int_aware_sort((ex_stride[i], i)), ex_stride[i], i) for i in range(dim)]
    heapq.heapify(pending)
    candidates = {}
    while pending:
        i_stride, i = heapq.heappop(pending)
        if i_stride in candidates:
            stride[i] = candidates[i_stride]
        else:
            stride[i] = i_stride
        candidates[i_stride * ex_size[i]] = InferStride(i)
```
I believe this is much simpler to read/understand.

In addition on a microbenchmark it seems significantly faster:
- 2x faster for the "common case" ( `(size=[3, 3, 3, 3], stride=[1, 3, 9, 27])`)
- 4x faster for a permutation over all 4d strides from 0 - 27

The only difference in output is that when there are multiple strides to pick from the two algorithms can pick different ones - for example for an input of:
```
(size=[13, 2, 26, 2, 0], stride=[17, 11, 22, 11, 27])
```
the original algorithm will output
```
[17, 11, InferStride(dim=1), 11, 27]
```
but this algorithm outputs
```
[17, 11, InferStride(dim=3), 11, 27]
```
I don't think that matters as long as the final values are internally consistent?

Reply:
And actually - it doesn't actually need the minheap since we're never inserting.  Changing it to just do a single sort of `pending` and then iterating over it improves the speed difference to 2.5x and 4.5x.